src/librustc_trans/assert_module_sources.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! This pass is only used for UNIT TESTS related to incremental
+//! compilation. It tests whether a particular `.o` file will be re-used
+//! from a previous compilation or whether it must be regenerated.
+//!
+//! The user adds annotations to the crate of the following form:
+//!
+//! ```
+//! #![rustc_partition_reused(module="spike", cfg="rpass2")]
+//! #![rustc_partition_translated(module="spike-x", cfg="rpass2")]
```

Comment:
> I just figured that there is the third case of a partition being neither re-used nor re-translated but removed entirely. 

I was anticipating a third-case when I did the names, but I'm not sure now if that makes sense. The third case I had in mind was one where we had saved the LLVM IR pre-optimization -- but in that case, I realize now you might still want `Translated`. Or at least the backend, optimizing code doesn't care where the LLVM IR came from (though we might use another variant just for unit testing).

I hadn't thought about the case of deleting a partition -- I don't know if we would need that for any reason? Just to cleanup the saved file, are you thinking? I had imagined we would do some sort of "GC" where, on startup--or maybe termination--we just crawl the incr-comp directory and delete random files that don't correspond to known work-products. In fact, I had such code, but I remember now I wound up deleting it because in the newer scheme I didn't *need* it to get things working.